ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,572 @@
+     * duplicate.
+     */
+    if (!side_effect_free) {
+        data->side_effect_free = false;
+        return DR_EMIT_DEFAULT;
+    }
+    data->side_effect_free = true;
+
+    /* Duplication process.
+     * Consider the following example bb:
+     *   instr1
+     *   instr2
+     *   term_instr
+     *
+     * In this stage, we just duplicate the bb (except for its terminating
+     * instruction) and add special labels to the original and duplicated blocks. The
+     * duplicated bb is for now skipped with a jump. The insert instrumentation stage will
+     * remove this jump, and add saving/restoring of machine state and the state
+     * comparison.
+     * Note that there might be no term_instr (no control transfer instruction) and the bb
+     * just falls-through. Even with no term_instr the jmp and the TERM label are inserted
+     * in the same way, as shown in this example.
+     *
+     * The example bb is transformed, in this stage, as follows:
+     * ORIG_BB:
+     *   instr1
+     *   instr2
+     *   jmp TERM
+     *
+     * COPY_BB:
+     *   instr1
+     *   instr2
+     * TERM:
+     *   term_instr
+     *
+     */
+
+    /* Create a clone of bb. */
+    instrlist_t *copy_bb = instrlist_clone(drcontext, bb);
+
+    /* Create and insert the labels. */
+    drstatecmp_insert_labels(drcontext, bb, instrlist_first(bb), DRSTATECMP_LABEL_ORIG_BB,
+                             /*preinsert=*/true);
+    instr_t *copy_bb_label =
+        drstatecmp_insert_labels(drcontext, copy_bb, instrlist_first(copy_bb),
+                                 DRSTATECMP_LABEL_COPY_BB, /*preinsert=*/true);
+    instr_t *term_inst_copy_bb = instrlist_last_app(copy_bb);
+    instr_t *term_label = NULL;
+    if (instr_is_cti(term_inst_copy_bb) || instr_is_return(term_inst_copy_bb)) {
+        term_label = drstatecmp_insert_labels(drcontext, copy_bb, term_inst_copy_bb,
+                                              DRSTATECMP_LABEL_TERM, /*preinsert=*/true);
+    } else {
+        term_label = drstatecmp_insert_labels(drcontext, copy_bb, term_inst_copy_bb,
+                                              DRSTATECMP_LABEL_TERM, /*preinsert=*/false);
+    }
+    /* Need an operand for the jmp instr targeting this label. */
+    opnd_t term_label_opnd = opnd_create_instr(term_label);
+
+    /* Replace the terminating instruction of the original bb with a jmp to the
+     * terminating instruction of the bb clone (if any).
+     */
+    instr_t *term_inst = instrlist_last_app(bb);
+    instr_t *jmp_term = XINST_CREATE_jump(drcontext, term_label_opnd);
+    if (instr_is_cti(term_inst) || instr_is_return(term_inst)) {
+        instrlist_replace(bb, term_inst, jmp_term);
+        instr_destroy(drcontext, term_inst);
+    } else {
+        /* If there is no control transfer, append a jmp at the end of the bb. */
+        instrlist_postinsert(bb, term_inst, jmp_term);
+    }
+
+    /* Append the instructions of the bb copy to the original bb. */
```

Comment:
But this will present the copy to all the instrumentation libraries and clients.  So drreg will consider all the reg uses in the copy when deciding which scratch regs to use, which will warp its behavior so it won't match what it would do w/o a copy.  IIRC its liveness analysis does consider jumps?  Is that why a jump was added?  Please add comments if so.  But its reg use counts will ignore the jump.  Plus other libraries/clients frequently count app instrs or take actions across app instrs: the most common is putting a clean call or batched action at the end of the block, which would mean inside this copy.  Then below you remove all the instrumentation: so the resulting block will not contain the instrumentation in the original part that would have been there.

I think this copy has to be hidden from the instrumentation, and only added in the insru2instru phase or something like that.

Reply:
@johnfxgalea -- Any thoughts on the best way to hide the extra copy from instrumentation?  It seems like either using the drbbdup approach with a drmgr-involved modified interface to hide the other copies, or adding it in the instru2instru phase (or I suppose insertion w/ very low priority) are the options: with the latter technically breaking rules on adding app instructions.

Reply:
Yep, the jump was added to allow drreg to do liveness analysis correctly but you are right that in many cases linear control flow is assumed and the copy will affect various instrumentation passes. To me doing everything in the instru2instru phase seems to be the best choice. Besides the main motivation for the comparison library is to check for clobbering by instrumentation. So, the safest option is to add the checks at the very end without influencing any other instrumentation. The code will also be simpler, since there is bb-level scope in instru2instru and everything will done in a single pass.

Reply:
instru2instru does seem simple, but I don't really like the idea of breaking drmgr's instrumentation contract with the addition of app instructions, especially if this becomes a common occurrence in the future. Then again, other solutions require more work.

If performance is not a major problem, an alternative approach might be to rely on re-direction, i.e., with the use of drbbdup,  first execute the instrumented case. Then rely on a clean call, which is placed at the end, to set a flag and redirect execution again to the same basic block. The active flag will instead lead drbbdup to dispatch to the uninstrumented case (which will only include instrumentation to reset the flag).

Reply:
Performance is not that critical for now (although it might be in the future if it proves useful and it is used beyond test apps), but code simplicity is. The point of this lib is to detect bugs in other DR components while it remains bug-free itself. If it is too complex or relies on other extensions (and thus inherit their complexities and potential bugs) then its usefulness decreases. Bugs in drstatecmp or its deps could lead to time-wasting false positives (incorrect detection of mismatch that needs to be investigated manually) or false negatives (undetected clobbering). I would also count as detrimental any intrusive effect of drstatecmp or its deps on the behavior of other DR components.

Reply:
All great points, and a big fan of simplicity.

> while it remains bug-free itself

The only issue I see is that going against the specification of drmgr is a bug in itself.

We could document the violating use of the instru2instru phase really well, stressing that DrStateCmp is just a debugging tool, but if we can find a better way to avoid this exception to the rule, that would be great! In general, exceptions to rules could potentially lead to nasty bugs or maintenance issues in the future, and this is something we should avoid.

Keeping in mind points on simplicity, how about we introduce a new final special drmgr phase (maybe call it `bb2debug`?), which is suited for DrStateCmp and other debugging extensions that we might create in the future? The phase will be the the last phase considered by drmgr and allows for flexible instrumentation, including app instr. Its introduction appears that it won't break drmgr rules.

Reply:
Having a fifth drmgr phase might be a good idea. The fourth phase is for instrumentation optimizations and there will be a fifth phase that is for instrumentation debugging. Seems appropriate. I am fine with more work as long as DrStateCmp itself remains simple.

Reply:
Yeah! If we do include the fifth stage, I'm not sure if we should document it solely for use by so called "debugging" extensions or allow other "normal" extensions to possibly register it as well. There could be some small differences in the implementation. If we go for phase 5 to be strictly for debugging tools, then we could potentially have an internal header file, and leave drmgr.h unchanged.

Maybe @derekbruening also has an opinion on this?

Reply:
My first reaction is to be hesitant to change the core phases: I started implementing a split of the insertion phase and ended up bailing because of complexities with drbbdup and other existing users of the current phases (filed aborted attempt as #4947).  But a completely separate phase that doesn't touch the existing key phases of app2app-analysis-insertion is probably going to have minimal disruption so it seems fine in that respect.

I kind of like the drbbdup re-execute idea at a high level: drbbdup is an existing framework supporting construction of two separate versions, which is exactly what we want here.  But having to do the full redirect to chain two versions, which drbbdup wasn't designed for, makes it much less appealing.

> If we go for phase 5 to be strictly for debugging tools, then we could potentially have an internal header file, and leave drmgr.h unchanged.

Could you elaborate on this?  I assumed we would want a public phase 5 since we want these debugging tools to be first-class libraries usable with new instrumentation schemes.

Reply:
Put simply, we need to decide whether we should have access to phase 5 via drmgr_priv.h or drmgr.h. If we decide that phase 5 is solely intended for debugging tools, maybe drmgr_priv.h would be sufficient. If we envisage phase 5 to be used more generally, then drmgr.h ?

Reply:
OK I could see separating, or documenting inside drmgr.h that it is meant for only certain uses.  But even if separated it would still be a published header using exported drmgr library functions, right?  So it wouldn't really be private/internal: just documented as for narrow use cases.

Reply:
But even if separated it would still be a published header using exported drmgr library functions, right? 

Yep, but generally users should include drmgr.h and not drmgr_priv.h.

Reply:
OK, though as a nit I would have a different name, since _private.h in DR means an interface not exported at all outside a module/library.